File: 2020/python/day5.py
from std import *
from functools import reduce

with open("../input/input5.txt", "r") as fd:    
    seats = sorted([reduce(lambda a, x: (a<<1)+(ord(x)%7)%2, line.strip(), 0) for line in fd])
    print("A:", seats[-1])
    print("B:", sum(range(seats[0], seats[-1]+1)) - sum(seats))

File: 2020/haskell/day5.hs
import Data.Char (ord)
import Data.List (sort)

main = do
    c <- readFile "../input/input5.txt"
    let seats = sort $ map toNum (lines c)
    putStrLn ("A: " ++ show (last seats))
    putStrLn ("B: " ++ show (sum [head seats .. last seats] - sum seats))
    where
        toNum = foldl (\a x -> 2*a + mod (mod (ord x) 7) 2) 0